src/util/JoinAlgorithms/JoinAlgorithms.h

```diff
@@ -264,11 +274,24 @@ CPP_template(typename Range1, typename Range2, typename LessThan,
         mergeWithUndefLeft(it, std::begin(left), it1);
       }
 
-      for (; it1 != endSame1; ++it1) {
-        checkCancellation();
-        cover(it1);
-        for (auto innerIt2 = it2; innerIt2 != endSame2; ++innerIt2) {
-          compatibleRowAction(it1, innerIt2);
+      // Add all the matching rows to the result.
+      // TODO<joka921> We should at some point enforce that all the
+      // `CompatibleRowAction`s support adding multiple rows at once.
```

Comment:
How many implementations are there that this isn't trivial to do?

Reply:
It is all the materialized versions of join/optional/minus with all their different mechanisms for undef and not + all unit tests. Given the deadline tomorrow I am not sure whether I should include this here.